libp2p/dialer.nim

```diff
@@ -81,16 +81,17 @@ proc dialAndUpgrade(
           if dialed.dir != dir:
             dialed.dir = dir
           await transport.upgrade(dialed, peerId)
+        except CancelledError as exc:
```

Comment:
Could you please elaborate on why this is necessary and why we don't need to call `close`?

Reply:
oops, close should stay (ed6be85d6e3d72008070625c0fb1637ceed37095) - the point was to not return `nil` which makes the dialler try the next address instead of aborting the dialling

Reply:
Could we catch only `LPError` instead? https://github.com/vacp2p/nim-libp2p/blob/c5db35d9b006b597886440293ff5cbd86162452c/libp2p/transports/transport.nim#L88

Reply:
we could, but for that to be safe, we would need to annotate dialAndUpgrade with raises annotations as well, so that changes in `upgrate` would be caught by the compiler - this would significantly increase the scope of this PR

Reply:
The code catches and raises `CancelledError`, I thought this would be equivalent to only catching `LPError`. Upgrade - the code in the `try/except` only raises those two errors. But thinking more about it, you said we still want to call `close`when catching `CancelledError`, so my suggestion probably doesn't improve anything. Btw, did you forget to call `close`?

Reply:
ah, highlighted wrong commit: 047019ae0fdd82630b148193c6b5284d042bc210 - close should be there

Reply:
> only raises those two errors.

if there is no raises annotation and the function being called starts raising a new exception, there's no way to tell which code is affected - therefore, when writing exception _handling_ code, it is reasonable to assume the worse case ("any exception can happen") even  if today there is an annotation. when we add `raises` here, we can catch more specific exceptions because changes in transport will be made visible

Reply:
Isn't it annotated in the code I mentioned?

Reply:
Ah, I see what you mean now, didn't think about it. So even if the called proc is annotated, it isn't safe to trust it if the caller isn't annotated as well.

Reply:
But thinking more about it, all errors in this project should inherit from `LPError`, shouldn't they? In theory, it should be fine then.

Reply:
Maybe it's even more correct as `LPError` should represent all the expected errors we thought about and are fine ignoring in this case. All the others shouldn't probably be swallowed here.

Reply:
Broadly, it's best when every abstraction has its own hierarchy - ie failure to decode a multihash is not related to a transport being closed in any meaningful way so having both derived from lperror doesn't really have any underlying motivation except that they happen to be implemented in the same codebase.

As such, it's usually best if exceptions are mapped to the abstraction layer that they're operating at, and each layer translates the exceptions coming from other layers to their own level - ie "socket closed" becomes "transport closed" becomes "peer disconnected" as it travels through the layers.

Reply:
as noted though, changing exception types is _usually_ a breaking change, so it needs to be done carefully - in this particular case, it needs to be done across the entire transport hierarchy at the same time.